android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java

```diff
@@ -38,134 +44,151 @@
+				} else {
+					Log.e(TAG, "AudioRecorder.start() failed to start recording. Reason: Unknown");
+				}
+			}
+		} catch (Exception ex) {
+			Log.e(TAG, "AudioRecorder.start() failed to start recording.", ex);
+		}
 
-		if (audioRecord.getState() == 1) {
-			audioRecord.setRecordPositionUpdateListener(onRecordPositionUpdateListener);
-			audioRecord.setPositionNotificationPeriod(bufferSize / 4);
-			audioRecord.startRecording();
-			audioRecord.read(audioData, 0, bufferSize);
-			recording = true;
+		// If we failed to start recording above, then clean-up any hanging resources.
+		if (!this.recording) {
+			stopRecording();
 		}
 	}
 
 	public String stopRecording()
 	{
+		// Stop recording and produce the WAV file.
 		File resultFile = null;
-		//Guard for calling stop before starting the recording
-		if (audioRecord != null) {
-			int recordState = audioRecord.getState();
-			if (recordState == 1) {
+		if (this.audioRecord != null) {
+			// Update state.
+			this.recording = false;
+			this.paused = false;
+			this.stopped = true;
+
+			// Stop recording.
+			int channelCount = this.audioRecord.getChannelCount();
+			if (this.audioRecord.getState() == AudioRecord.STATE_INITIALIZED) {
 				audioRecord.stop();
 			}
-			audioRecord.setRecordPositionUpdateListener(null);
-			audioRecord.release();
-			audioRecord = null;
+			this.audioRecord.setRecordPositionUpdateListener(null);
+			this.audioRecord.release();
+			this.audioRecord = null;
+
+			// Update recorded WAV file's header info.
+			try {
+				if (this.randomAccessFile != null) {
+					this.randomAccessFile.seek(0);
+					long totalFileBytes = this.randomAccessFile.length();
+					writeWaveFileHeader(
+						this.randomAccessFile,
+						totalFileBytes - WAV_FILE_HEADER_TOTAL_BYTE_COUNT,
+						totalFileBytes - WAV_FILE_HEADER_RIFF_SUBCHUNK_BYTE_COUNT,
+						RECORDER_SAMPLE_RATE,
+						channelCount,
+						RECORDER_BPP * RECORDER_SAMPLE_RATE * channelCount / 8);
+					resultFile = this.tempFileReference;
+				}
+			} catch (Exception ex) {
+				Log.e(TAG, "AudioRecorder.stop() failed to write audio file.", ex);
+			}
+		}
+
+		// Close the output file.
+		if (this.randomAccessFile != null) {
 			try {
-				resultFile = TiFileHelper.getInstance().getTempFile(AUDIO_RECORDER_FILE_EXT_WAV, true);
-				createWaveFile(resultFile.getAbsolutePath());
-			} catch (IOException e) {
-				e.printStackTrace();
+				this.randomAccessFile.close();
+			} catch (Exception ex) {
+				ex.printStackTrace();
 			}
+			this.randomAccessFile = null;
 		}
-		return resultFile != null ? resultFile.getAbsolutePath() : "";
+
+		// Return the file path if successfully recorded to file.
+		return (resultFile != null) ? resultFile.getAbsolutePath() : null;
 	}
 
 	public void pauseRecording()
 	{
-		//Guard for calling pause before starting the recording
-		if (audioRecord != null) {
-			paused = true;
-			audioRecord.stop();
+		if (this.audioRecord != null) {
+			this.audioRecord.stop();
```

Comment:
Yes, we need the `null` checks. We set the `audioRecord` member variable to `null` when the `stopRecording()` method gets called. We create a new instance when the `startRecording()` method gets called.

Reply:
Although `this.audioRecord.stop()` can be called multiple times, it might be good to do this:
```Java
if (!this.paused && this.audioRecord != null) {
```

Reply:
An even better solution would be to ditch those 3 boolean member variables (which I agree is silly) and fetch the state from the audio recorder object itself similar to the below. Although I've already confirmed that there is no harm in calling the native `AudioRecord.start()` if already started and `AudioRecord.stop()` if already stopped.
```java
if (this.audioRecord == null) {
	// Stopped.
} else if (this.audioRecord.getRecordingState() == AudioRecord.RECORDSTATE_RECORDING) {
	// Recording.
} else {
	// Paused.
}
```